setup.py

```diff
@@ -54,21 +125,35 @@ def set_builtin(name, value):
 
 # Custom builder to handler compiler flags. Edit if needed.
 class build_ext_subclass(build_ext):
+
+    def _omp_compile_link_args(self, compiler):
+        # Get OpenMP setting from environment
+        use_omp = OMP_SETTING_TABLE[os.environ.get('USE_OMP', 'probe')]
+        # OpenMP is not supported with default clang
+        # Conda provides its own compiler which does support openmp
+        if use_omp in {'probe', 'omp'} and is_macOS() and not is_conda_interpreter():
+            macos_compile_args, macos_link_args = macOS_omp_options_from_probe()
+            if macos_compile_args is not None and macos_link_args is not None:
+                # we found a libomp that we should be able to use
+                use_omp = 'omp'
+                return use_omp, macos_compile_args + OMP_COMPILE_ARGS[use_omp], macos_link_args + OMP_LINK_ARGS[use_omp]
+            else:
+                # likely default system clang compiler with no libomp in typical locations
+                use_omp = None
+        if use_omp == 'probe':
+            use_omp = 'msvc' if compiler == 'msvc' else 'gomp'
+        return use_omp, OMP_COMPILE_ARGS.get(use_omp, []), OMP_LINK_ARGS.get(use_omp, [])
+
     def build_extensions(self):
         comp = self.compiler.compiler_type
+        use_omp, omp_comp, omp_link = self._omp_compile_link_args(comp)
```

Comment:
Besides the print statement `use_omp` isn't used anymore, right? It makes me wonder if those `if` statements in the above helper method can be simplified even further. Perhaps getting `OMP_COMPILE_ARGS.get(use_omp, [])` earlier in the method and then adding `macos_compile_args` later on if needed...hhmmm except you might force `omp` later. Not sure.

The print could maybe be moved up into the helper or removed? Thoughts?

Reply:
The print only shows up if you add the -v to the pip, currently; if there's a better way to have an "info" message during a setup.py, that would be preferred. Having it as optional debug is useful, to show concisely what choices have been taken. Placing it inside the _omp_compile_link_args would require a refactoring to avoid a DRY violation. Might be straying a bit too far into the white glove treatment, in other words. 

I have more concerns about making sure Windows builds still work, and edge cases like "I have a linux distro that's clang-focused" (OpenMandriva?) have a path to build without having to edit the setup.py. I think that's the case, given I can swap between gcc/gomp and clang/omp on macos successfully by just changing environment values.

Reply:
> Placing it inside the _omp_compile_link_args would require a refactoring to avoid a DRY violation.

Why? Because there are two return statements? My instinct is to think any refactoring down those lines would only make the code cleaner. For example, right now you set `use_omp = None` in an `else` statement, but that has no usage or meaning anywhere else in the code. If you wrapped the mac stuff into a new function *or* had `macOS_omp_options_from_probe` return empty lists instead of `None` then the rest of the function/if statements just goes away. Or at least you could do something like:

```python
compile_args = []
link_args = []
...

macos_compile_args, macos_link_args = macOS_omp_options_from_probe()
if macos_compile_args and macos_link_args:
    compile_args += macos_compile_args
    link_args += macos_link_args

...

```

Hm maybe this doesn't clean it up this much. I completely didn't include the OMP_COMPILE_ARGS. I think there is a slightly cleaner way to structure it, but maybe it isn't worth the effort.

Reply:
Assuming MSVC can handle loading openmp by default as you have it, what if it was split up like this:

```
if use_omp == "probe":
    if is_macOS() and not is_conda_interpreter():
        use_omp = "omp"
    elif compiler == "msvc":
        use_omp = "msvc"
    else:
        use_omp = "gomp"
    print(f"Probing OpenMP support will attempt to use {use_omp}")
compile_args = []
link_args = []
if use_omp == "omp" and is_macOS():
    extra_compile_args, extra_link_args = macOS_omp_options_from_probe()
    if not extra_compile_args:
        print("Could not find OpenMP on MacOS")
    compile_args += extra_compile_args
    link_args += extra_link_args
compile_args += OMP_COMPILE_ARGS.get(use_omp, [])
link_args += OMP_LINK_ARGS.get(use_comp, [])
```

Put in to two separate helper functions this clears up "guess OpenMP usage based on platform/compiler" and "collect all compile and link args for OpenMP".